static/app/components/arithmeticBuilder/tokenizer.tsx

```diff
@@ -0,0 +1,227 @@
+          line: 1,
+          column: 1,
+        },
+    end: next
+      ? next.start
+      : prev
+        ? prev.end
+        : {
+            offset: 0,
+            line: 1,
+            column: 1,
+          },
+  };
+  return new TokenFreeText(location, '');
+}
+
+function tryTokenizeExpression(expression: string): Token[] {
+  if (!expression.trim()) {
+    return [];
+  }
+
+  const tc = new TokenConverter();
+  return grammar.parse(expression, {tc});
+}
+
+export function tokenizeExpression(
+  expression: string,
+  options?: {intersperseSpace: boolean}
+): Token[] {
+  const {intersperseSpace = true} = options || {};
+
+  if (!intersperseSpace) {
+    return tryTokenizeExpression(expression);
+  }
+
+  let loc: LocationRange | null = null;
+
+  const tokens: Token[] = [];
+
+  for (const token of tryTokenizeExpression(expression)) {
+    const prev = tokens[tokens.length - 1];
+    if (isTokenFreeText(prev) && isTokenFreeText(token)) {
+      prev.merge(token);
+    } else {
+      // make sure to inject a free text token between every pair of non free space
+      // tokens to allow users to enter things between them
+      if (!isTokenFreeText(prev) && !isTokenFreeText(token)) {
+        tokens.push(space(loc, token.location));
+      }
+
+      tokens.push(token);
+      loc = token.location;
+    }
+  }
+
+  // make sure to check if we need a space at the end
+  if (tokens.length <= 0 || !isTokenFreeText(tokens[tokens.length - 1])) {
+    tokens.push(space(loc, null));
+  }
+
+  const counters: Record<TokenKind, number> = {
+    [TokenKind.UNKNOWN]: 0,
+    [TokenKind.PARENTHESIS]: 0,
+    [TokenKind.OPERATOR]: 0,
+    [TokenKind.FREE_TEXT]: 0,
+    [TokenKind.ATTRIBUTE]: 0,
+    [TokenKind.FUNCTION]: 0,
+  };
+
+  // assign an unique key to each token based on it's type
+  // and it's position in the list
+  for (const token of tokens) {
+    const i = counters[token.kind]++;
+    token.key = makeTokenKey(token.kind, i);
+
+    if (isTokenFunction(token)) {
+      for (const attr of token.attributes) {
+        const j = counters[attr.kind]++;
+        attr.key = makeTokenKey(attr.kind, j);
+      }
+    }
+  }
+
+  return tokens;
+}
+
+export function makeTokenKey(kind: TokenKind, i: number = 0): string {
+  return `${kind}:${i}`;
+}
+
+function parseTokenKey(key: string) {
+  const [kind, indexStr] = key.split(':');
+  const index = parseInt(indexStr!, 10);
+  return {kind: toTokenKind(kind!), index};
+}
+
+function isTokenKeyOfKind(key: string, kind: TokenKind) {
+  return key.startsWith(kind);
+}
+
+export function nextSimilarTokenKey(key: string): string {
+  const {kind, index} = parseTokenKey(key);
+  return makeTokenKey(kind, index + 1);
+}
+
+export function nextTokenKeyOfKind(
+  state: ListState<Token>,
+  token: Token,
+  kind: TokenKind
+): string {
+  let key: string | null = null;
+
+  for (const tokenKey of state.collection.getKeys()) {
+    if (tokenKey === token.key) {
+      break;
+    }
```

Comment:
Doesn't look like this function is being used yet, so I'm not sure what order these tokes exist in this state, but is the break expected to be here/work? will we break early?

Reply:
The tokens exist in the order of the input. So the intention is to find a the last token of the matching type before the specified token and based on that we can determine what the next token should be.